worker/rhpv3.go

```diff
@@ -376,117 +373,113 @@ func (w *worker) initTransportPool() {
 func (a *accounts) ForHost(hk types.PublicKey) *account {
 	accountID := rhpv3.Account(a.deriveAccountKey(hk).PublicKey())
 	return &account{
-		as:          a.as,
-		id:          accountID,
-		key:         a.key,
-		host:        hk,
-		shutdownCtx: a.shutdownCtx,
+		as:   a.as,
+		id:   accountID,
+		key:  a.key,
+		host: hk,
 	}
 }
 
-// WithDeposit increases the balance of an account by the amount returned by
-// amtFn if amtFn doesn't return an error.
-func (a *account) WithDeposit(ctx context.Context, amtFn func() (types.Currency, error)) error {
-	_, lockID, err := a.as.LockAccount(ctx, a.id, a.host, false, accountLockingDuration)
+func withAccountLock(ctx context.Context, as AccountStore, id rhpv3.Account, hk types.PublicKey, exclusive bool, fn func(a api.Account) error) error {
+	acc, lockID, err := as.LockAccount(ctx, id, hk, exclusive, accountLockingDuration)
 	if err != nil {
 		return err
 	}
+
 	defer func() {
-		unlockCtx, cancel := context.WithTimeout(a.shutdownCtx, 10*time.Second)
-		a.as.UnlockAccount(unlockCtx, a.id, lockID)
-		cancel()
+		select {
+		case <-ctx.Done():
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(context.Background(), time.Minute)
```

Comment:
This should always be the case and not just on `case <-ctx.Done()`.
I also don't think this needs to be deferred anymore